### PR TITLE
msg/async: fix crash that writing char to nonblock-fd gets EAGAIN in EventCenter::wakeup

### PR DESCRIPTION
**Pull request by Liuchang0812**
Signed-off-by: liuchang0812 <liuchang0812@gmail.com>

@yuyuyu101 take a look, please

### CODE REVIEW

**Comment by yuyuyu101** [comment]
EWOULDBLOCK is the same value as EAGAIN

**Comment by Liuchang0812** [comment]
linux man says:

       EAGAIN or EWOULDBLOCK
              The file descriptor fd refers to a socket and has been marked
nonblocking (O_NONBLOCK), and the read would block.  POSIX.1-2001 allows
either error to be returned for  this  case,  and
              does not require these constants to have the same value, *so
a portable application should check for both possibilities.*

IMO, It's better to check EWOULDBLOCK

2017-03-07 14:55 GMT+08:00 Haomai Wang <notifications@github.com>:

**Comment by Liuchang0812** [comment]
Change Log:
1. remove EWOULDBLOCK check
2. rebase and repush commit

@yuyuyu101 thanks

**Comment by yuyuyu101** [comment]
hmm, I don't think it will return EAGAIN for eventfd... if did, it should be bug....

**Comment by Liuchang0812** [comment]
@yuyuyu101 We benchmark H-release with async-messager, and get one core-dump that assert failed at https://github.com/ceph/ceph/blob/hammer/src/msg/async/Event.cc#L246

IMO, nonblocking pipe's writing would gets EAGAIN.

**Comment by Liuchang0812** [comment]
some log that we added for this

```
Mar  3 20:38:14  bash: write failed: -1 errno 11 msg/async/Event.cc: In function 'void EventCenter::wakeup()' thread 7
f1d64b89700 time 2017-03-03 20:38:14.861197
Mar  3 20:38:14  bash: msg/async/Event.cc: 278: FAILED assert(n == 1)
```

**Comment by yuyuyu101** [comment]
hammer? I don't think hammer's issue will apply to master... They're totally different... Do you backport master's async msgr to hammer or?

**Comment by Liuchang0812** [comment]
@yuyuyu101 I have read the code on master, there are same problem seems. It would be better to backport all changes to Hammer.

you added EAGAIN check in read pipe in this commit https://github.com/ceph/ceph/commit/f62f28ceea5649dc821365348093f8541db66e88

**Comment by Liuchang0812** [comment]
@yuyuyu101 sorry, release is Jewel

**Comment by yuyuyu101** [comment]
@Liuchang0812 read is different to write, it make sense that we would like to get EAGAIN from reading. but writing must be successful. Otherwise it means potential thread wakeup leaking risk.

**Comment by yuyuyu101** [comment]
do you mean you still meet the crash in master branch?

**Comment by Liuchang0812** [comment]
no, I don't test it on master code.

**Comment by yuyuyu101** [comment]
Jewel release exists already known issues about async msgr. Since K release, we make it default.

**Comment by Liuchang0812** [comment]
thanks for your explaining.  It seems that `wakeup` logical is same between Jewel and Master branch. We should take a deep look and try to fix this problem. 

Here is libev's evpipe impletation:
https://github.com/enki/libev/blob/93823e6ca699df195a6c7b8bfa6006ec40ee0003/ev.c#L2466

**Comment by yuyuyu101** [comment]
... AFAK I haven't see any bug report related to this in stable version. If met, it must be a bug we need to handle instead of ignore this.

**Comment by Liuchang0812** [comment]
ok, do you have any suggestion about jewel release, We have to use Jewel now.  appreciated!

**Comment by Liuchang0812** [comment]
@tchaikov take a look please?

**Comment by yuyuyu101** [comment]
we don't have plan to backport now. why not simple msgr until L release?

**Comment by Liuchang0812** [comment]
We have too many OSDs, simple msgr will issue pthread_create failed. :rofl:

**Comment by yuyuyu101** [comment]
@Liuchang0812 what about upgrading to K?

**Comment by Liuchang0812** [comment]
@yuyuyu101 i will have a try, need to discussed with others

**Comment by Liuchang0812** [comment]
@yuyuyu101 I found the reason. We use epool ET mode on pipe, but async msgr read 256 chars one-time. so we will not read pipe forever. Need we backport related patches to Jewel.

**Comment by yuyuyu101** [comment]
ok, but keep in mine.. it still exists several problems...

**Comment by Liuchang0812** [comment]
but i still think there is possibility to write pipe full before reading
pipe. it would be better to add EAGAIN check to avoid crash.

2017-03-08 17:48 GMT+08:00 Haomai Wang <notifications@github.com>:

File: src/msg/async/Event.cc [hunk 1]
```diff
@@ -323,8 +323,11 @@ void EventCenter::wakeup()
   // wake up "event_wait"
   int n = write(notify_send_fd, &buf, sizeof(buf));
   if (n < 0) {
-    ldout(cct, 1) << __func__ << " write notify pipe failed: " << cpp_strerror(errno) << dendl;
-    ceph_abort();
+    if (errno != EAGAIN)
+    {
```
Comment (yuyuyu101) [inline src/msg/async/Event.cc]: make this if () {
}

style

**Comment by Liuchang0812** [comment]
Changelog:

1. corrected code style
2. rebased and repushed


@yuyuyu101 I would like to back-port some commits to Jewel, is it ok?

**Comment by yuyuyu101** [comment]
@Liuchang0812 of course

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2017-03-16_11:38:52-rados-wip-kefu-testing---basic-mira/
- http://pulpito.ceph.com/kchai-2017-03-17_04:29:25-rados-wip-kefu-testing---basic-mira/

failed tests tracked by http://tracker.ceph.com/issues/19312

**Pull request merged**